### PR TITLE
Azeey/marsupial

### PR DESCRIPTION
**Pull request by nkoenig**
This is a copy of the contents in #396 with the addition of drone platform model.

The following is from #396.

This adds a command line argument `marsupial#:=<parentRobotName>:<childRobotName>`  where `parentRobotName` and `childRobotName` are names of robots specified using `robotName#`. The variable `MARSUPIAL_VALID_PARENT_ROBOT_TYPES` is used to store valid parent robot types, which is currently just the `X1` . Similarly, `MARSUPIAL_VALID_CHILD_ROBOT_TYPES` is used to store valid child robot types, which is currently just the `X3`.

It works by adding the `DetachableJoint` system to the parent robot. This is done by parsing the XML string generated by the `spawn` functions and injecting the `<plugin>` tag for the `DetachableJoint`.

This PR adds a new topic, <childRobotName>/detach`, which is used for detaching the child robot.

## Example
```
ign launch -v 4 cave_circuit.ign  robotName1:=Xa robotConfig1:=X1_SENSOR_CONFIG_1 robotName2:=Xb robotConfig2:=X3_SENSOR_CONFIG_1 robotName3:=Xc robotConfig3:=X1_SENSOR_CONFIG_1 robotName4:=Xd robotConfig4:=X3_SENSOR_CONFIG_1 marsupial1:=Xa:Xb marsupial2:=Xc:Xd
```
To detach `Xb`, run:

```rostopic pub /Xb/detach std_msgs/Empty "{}"```

## Known Issues

~~There is currently a bug that seems to disable collisions between detached robots and the rest of the environment~~ See [ign-physics PR 31](https://github.com/ignitionrobotics/ign-physics/pull/31)

### CODE REVIEW

**Comment by azeey** [comment]
I ran
```
ign launch -v 4 cave_circuit.ign  robotName1:=Xa robotConfig1:=X1_SENSOR_CONFIG_1 robotName2:=Xb robotConfig2:=X3_SENSOR_CONFIG_1 robotName3:=Xc robotConfig3:=X1_SENSOR_CONFIG_1 robotName4:=Xd robotConfig4:=X3_SENSOR_CONFIG_1 robotName5:=Xe robotConfig5:=X1_SENSOR_CONFIG_2 robotName6:=Xf robotConfig6:=X3_SENSOR_CONFIG_2 marsupial1:=Xa:Xb  marsupial2:=Xc:Xd marsupial3:=Xe:Xf
```
and it looks like the position of the UAVs on the platform is not constant. 
image

Looking closer, I think the platform itself is lower in some of the robots.
image

**Comment by nkoenig** [comment]
Can you try again with: 4889d0f, and also get the latest version of: https://app.ignitionrobotics.org/OpenRobotics/fuel/models/DronePlatformX1

**Comment by azeey** [comment]
Same issue. The problem is that the X1 falls a small amount as soon as it's spawned. This is the exaggerated version. Not sure what we can do about it though
ezgif com-optimize (1)

**Comment by nkoenig** [comment]
I see. I'll have to create new models for each of the X1 models.

**Comment by nkoenig** [comment]
I've update this PR to spawn all vehicles at once. This now depends on:

1. https://github.com/ignitionrobotics/ign-msgs/pull/58
2. https://github.com/ignitionrobotics/ign-gazebo/pull/146
3. https://github.com/ignitionrobotics/ign-launch/pull/30

File: subt_ign/launch/cave_circuit.ign [hunk 1]
```diff
@@ -1191,27 +1386,95 @@
...
+             </include>
+           </sdf>
+         </spawn>
+       HEREDOC
+
+       # Wrap the temporary spawn string with a root element. This is needed
+       # because the xml parser needs a root element. When we're adding the
+       # plugin, we print only the contents of the root element.
+       spawnXml = REXML::Document.new("<ignition>#{spawnStringTmp}</ignition>")
+       search = "//spawn[@name='#{name}']"
```
Comment (azeey) [inline subt_ign/launch/cave_circuit.ign:1458]: Submitted models don't use the new `<spawn>` tag. Can we also check for the `ignition::launch::GazeboFactory` here?

Comment (azeey) [reply subt_ign/launch/cave_circuit.ign:1460]: Actually, we probably need to convert them to `<spawn>` tags either in the `spawner.rb` scripts or by parsing the result of the `spawner` function on line 1402.

File: subt_ign/launch/competition.ign [hunk 2]
```diff
@@ -42,13 +91,80 @@
...
+
+      marsupialParents[parent] = child
+    end
+  end
 %>
 
 <%
+  # Check if worldName is not defined or is empty/nil
+  if !defined?(worldName) || worldName == nil || worldName.empty?
+    $worldName = 'simple_cave_01'
```
Comment (azeey) [inline subt_ign/launch/competition.ign:163]: Should this be tunnel_qual?

File: subt_ign/launch/competition.ign [hunk 3]
```diff
@@ -150,104 +285,191 @@
...
-    <artifact>
-      <name>rescue_randy_3</name>
-      <type>TYPE_RESCUE_RANDY</type>
-    </artifact>
-    <artifact>
-      <name>rescue_randy_4</name>
-      <type>TYPE_RESCUE_RANDY</type>
-    </artifact>
-  </plugin>
+      <artifact>
```
Comment (azeey) [inline subt_ign/launch/competition.ign:300]: I am no up to date on which circuit/world has which artifacts. With these changes, it looks like they would all have the same artifacts. Is that okay?

File: subt_ign/launch/cave_circuit.ign (hunk 1 above)
Comment (nkoenig) [reply subt_ign/launch/cave_circuit.ign:1460]: I think I converted the submitted models in: f13f1a12c620541d5c8c0acb1bde18fe1755bd7d

File: subt_ign/launch/competition.ign (hunk 3 above)
Comment (nkoenig) [reply subt_ign/launch/competition.ign:300]: It's okay to list them here. The artifact list in the launch files capture all the possible artifacts. The world SDF files contain the artifacts for each world.

File: subt_ign/launch/competition.ign (hunk 2 above)
Comment (nkoenig) [reply subt_ign/launch/competition.ign:163]: It's okay to update this so that it points to the most recent circuit.

**Pull request merged**